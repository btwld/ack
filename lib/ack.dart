library;

export 'src/ack.dart';
