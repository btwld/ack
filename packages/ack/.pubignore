# Development and IDE files
melos_ack.iml
.claude/

# Build artifacts
build/
coverage/

# Local example (if any development examples exist)
example/
