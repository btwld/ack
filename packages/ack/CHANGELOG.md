## 0.2.0-beta.1 (2025-05-03)

* See [release notes](https://github.com/btwld/ack/releases/tag/v0.2.0-beta.1) for details.


## 0.2.0-beta.1

 - Bump "ack" to `0.2.0-beta.1`.

## 0.2.0

 - Bump "ack" to 0.2.0 with improved SchemaModel API and enhanced string validation

## 0.2.0

### Breaking Changes

- **SchemaModel API**: Completely redesigned the SchemaModel API for a more intuitive and streamlined experience
  - Changed constructor parameter from `Map<String, dynamic>` to `Object?` for better flexibility
  - Added automatic validation during construction
  - Added `isValid` getter and `getErrors()` method for easier error handling
  - Added abstract `getSchema()` method to define the schema
  - Removed `parse()`, `fromValidated()`, `validateMap()`, `validate()`, `createFromMap()`, and `initialize()` methods

### Improvements

- Improved error handling with more detailed error messages
- Better type safety with `Object?` instead of `dynamic`
- Simplified API with fewer methods and more intuitive usage
- **String Validation**: Enhanced string validation with improved methods
  - Changed `matches()` to validate full string patterns (anchored with ^ and $)
  - Added `contains()` method for partial string matching

## 0.1.2

- Improved error messages

## 0.1.1

-  Added deprecations

## 0.1.0

-  Error and exception improvements
-  Reworked validation and constraint workflows
-  Improved testing

## 0.0.2

-  Better JSON response parsing for OpenApiSchemaConverter

## 0.0.1

-  Initial version.