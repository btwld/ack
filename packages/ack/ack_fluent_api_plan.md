# Ack Fluent API Refactor Plan

This document outlines the plan to implement a fluent, Zod-inspired API for the Ack validation library. The work is broken down by schema type and is based on a detailed analysis of our current architecture and the capabilities of the Zod library.

---

### **Guiding Principles**

1.  **Universal Methods go on the Base Class**: Any capability that can apply to *any* schema (like making it nullable or adding a description) should be a method on the base `AckSchema`.
2.  **Specific Methods go in Extensions**: Capabilities that only make sense for a specific type (e.g., email validation for strings) should be implemented as extension methods on that specific schema class (`StringSchema`, `ListSchema`, etc.). This keeps the core classes clean.
3.  **Unsupported Features**: We will clearly mark features from Zod that we cannot or should not implement right now due to significant architectural differences (e.g., transformations, async validation).

---

### **1. `AckSchema` (Base Class Methods)**

These methods will be added directly to the `AckSchema<T>` class because they are universal.

| Method Signature | Zod Equivalent | Why Here? | Implementation Notes |
| :--- | :--- | :--- | :--- |
| `nullable()` | `.nullable()` | Makes any schema type nullable. This is the most fundamental and universal operation. | Already implemented via the `.nullable()` extension. No new work needed. |
| `default(T value)` | `.default(value)` | Sets a default value if the input is `null` or `undefined`. A universal concept. | Already implemented via the `defaultValue` constructor parameter. The `copyWith` pattern allows us to set it fluently. |
| `description(String text)` | `.describe(text)` | Adds a description for documentation/error message purposes. Universal. | Already implemented via the `description` constructor parameter. |

**Conclusion:** The most fundamental universal methods are already supported by our new architecture.

---

### **2. `StringSchema` Extensions**

These will be implemented in a new file: `lib/src/schemas/extensions/string_schema_extensions.dart`.

| Method Signature | Zod Equivalent | Why Here? | Support Level |
| :--- | :--- | :--- | :--- |
| `minLength(int n)` | `.min(n)` | String-specific length validation. | **Can Support Now** |
| `maxLength(int n)` | `.max(n)` | String-specific length validation. | **Can Support Now** |
| `length(int n)` | `.length(n)` | String-specific length validation. | **Can Support Now** |
| `email()` | `.email()` | String-specific format validation. | **Can Support Now** |
| `url()` | `.url()` | String-specific format validation. | **Can Support Now** |
| `uuid()` | `.uuid()` | String-specific format validation. | **Can Support Now** |
| `cuid()` | `.cuid()` | String-specific format validation. | **Can Support Now** (less common, maybe lower priority) |
| `datetime()` | `.datetime()` | String-specific format validation. | **Can Support Now** |
| `regex(RegExp re)` | `.regex(re)` | String-specific pattern matching. | **Can Support Now** |
| `startsWith(String s)` | `.startsWith(s)` | String-specific content validation. | **Can Support Now** |
| `endsWith(String s)` | `.endsWith(s)` | String-specific content validation. | **Can Support Now** |
| `trim()` | `.trim()` | **Transformation**. | **Cannot Support Now**. Our architecture validates, it doesn't transform the input value. |
| `toLowerCase()` | `.toLowerCase()` | **Transformation**. | **Cannot Support Now**. |
| `toUpperCase()` | `.toUpperCase()` | **Transformation**. | **Cannot Support Now**. |

---

### **3. `IntegerSchema` & `DoubleSchema` Extensions**

These will be in `lib/src/schemas/extensions/num_schema_extensions.dart`.

| Method Signature | Zod Equivalent | Why Here? | Support Level |
| :--- | :--- | :--- | :--- |
| `gt(num n)` / `greaterThan(num n)` | `.gt(n)` | Numeric comparison. | **Can Support Now** |
| `gte(num n)` / `min(num n)` | `.gte(n)` | Numeric comparison. | **Can Support Now** |
| `lt(num n)` / `lessThan(num n)` | `.lt(n)` | Numeric comparison. | **Can Support Now** |
| `lte(num n)` / `max(num n)` | `.lte(n)` | Numeric comparison. | **Can Support Now** |
| `isPositive()` | `.positive()` | Numeric sign check. | **Can Support Now** |
| `isNegative()` | `.negative()` | Numeric sign check. | **Can Support Now** |
| `isNonPositive()` | `.nonpositive()` | Numeric sign check. | **Can Support Now** |
| `isNonNegative()` | `.nonnegative()` | Numeric sign check. | **Can Support Now** |
| `multipleOf(num n)` / `step(num n)` | `.multipleOf(n)` | Numeric divisibility check. | **Can Support Now** |
| `finite()` | `.finite()` | Checks if a number is finite. | **Can Support Now** (More relevant for `DoubleSchema`) |
| `safe()` | `.safe()` | Checks if an integer is a "safe" integer. | **Can Support Now** (For `IntegerSchema`) |

---

### **4. `ListSchema` Extensions**

These will be in `lib/src/schemas/extensions/list_schema_extensions.dart`.

| Method Signature | Zod Equivalent | Why Here? | Support Level |
| :--- | :--- | :--- | :--- |
| `minLength(int n)` | `.min(n)` | List-specific length validation. | **Can Support Now** |
| `maxLength(int n)` | `.max(n)` | List-specific length validation. | **Can Support Now** |
| `length(int n)` | `.length(n)` | List-specific length validation. | **Can Support Now** |
| `nonempty()` | `.nonempty()` | A convenient shorthand for `.minLength(1)`. | **Can Support Now** |

---

### **5. `ObjectSchema` Extensions**

This is where things get more complex. Our `ObjectSchema` is powerful but doesn't have all of Zod's dynamic capabilities.

| Method Signature | Zod Equivalent | Why Here? | Support Level |
| :--- | :--- | :--- | :--- |
| `pick(List<String> keys)` | `.pick()` | **Structural Transformation**. | **Cannot Support Now**. This creates a *new* schema type by altering the properties. Our `copyWith` can't do this. |
| `omit(List<String> keys)` | `.omit()` | **Structural Transformation**. | **Cannot Support Now**. Same reason as `pick`. |
| `partial()` | `.partial()` | **Structural Transformation**. | **Cannot Support Now**. This changes the `required` status of all properties, which would require deep schema cloning. |
| `deepPartial()` | `.deepPartial()` | **Structural Transformation**. | **Cannot Support Now**. Even more complex than `partial`. |
| `merge(ObjectSchema other)` | `.merge()` | Combines two object schemas. | **Can Support Now (with limitations)**. We can implement a version of this that merges properties and `required` fields. |
| `passthrough()` | `.`passthrough()` | Allows unknown keys. | **Already Supported**. This is our `allowAdditionalProperties: true` flag. We could add a fluent method for it. |
| `strict()` | `.strict()` | Disallows unknown keys. | **Already Supported**. This is our `allowAdditionalProperties: false` flag. We could add a fluent method for it. |
| `catchall(AckSchema valueSchema)` | `.catchall()` | Validates all unknown keys against a schema. | **Cannot Support Now**. Our current implementation only checks for the *presence* of additional properties, not their *values*. |

---

### **6. Features We Cannot Support (The "Why")**

*   **Transformations (`.trim()`, `.toLowerCase()`, `.transform()`, `.refine()`)**:
    *   **Reason:** Our architecture has a clear separation of concerns: a schema's job is to **validate** an input and return a `SchemaResult` with either the original (but correctly typed) value or an error. It does not **mutate or transform** the input value. Zod's `.transform()` and `.refine()` fundamentally change the output value, which is a different paradigm. Supporting this would require a major redesign of the `parseAndValidate` flow.

*   **Structural Transformations (`.pick()`, `.omit()`, `.partial()`)**:
    *   **Reason:** These methods in Zod dynamically create *new* schema definitions at runtime by introspecting and cloning existing schemas. Our `copyWith` pattern is great for creating a modified *copy*, but it's not designed for this level of dynamic, structural alteration (e.g., removing keys from the `properties` map and the `requiredProperties` list simultaneously). It's possible, but would require significantly more complex logic in the `ObjectSchema`'s `copyWith` method.

*   **Asynchronous Validation (`.refineAsync()`)**:
    *   **Reason:** Our entire validation pipeline (`validate` -> `parseAndValidate` -> `SchemaResult`) is synchronous. Introducing `async` validation would require making the entire pipeline return `Future<SchemaResult>`, which is a massive and breaking change across the entire library. 