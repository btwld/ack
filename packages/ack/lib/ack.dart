/// ACK - A Comprehensive Kotlin-inspired validation library for Dart
library;

// Constraints
export 'src/constraints/constraint.dart';
// Context
export 'src/context.dart';
export 'src/schemas/extensions/list_schema_extensions.dart';
export 'src/schemas/extensions/numeric_extensions.dart';
export 'src/schemas/extensions/object_schema_extensions.dart';
// Extensions
export 'src/schemas/extensions/string_schema_extensions.dart';
// Core schemas
export 'src/schemas/schema.dart';
export 'src/validation/ack_exception.dart';
export 'src/validation/schema_error.dart';
// Validation results
export 'src/validation/schema_result.dart';
