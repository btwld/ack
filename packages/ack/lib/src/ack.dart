import 'schemas/extensions/string_schema_extensions.dart';
import 'schemas/schema.dart';

/// The main entry point for creating schemas with the Ack validation library.
///
/// Provides a fluent API for creating various schema types.
final class Ack {
  /// Creates a string schema.
  static StringSchema string() => const StringSchema();

  /// Creates a literal string schema that only accepts the exact [value].
  /// Similar to Zod's `z.literal("value")`.
  static StringSchema literal(String value) => string().literal(value);

  /// Creates an integer schema.
  static IntegerSchema integer() => const IntegerSchema();

  /// Creates a double schema.
  static DoubleSchema double() => const DoubleSchema();

  /// Creates a boolean schema.
  static BooleanSchema boolean() => const BooleanSchema();

  /// Creates an object schema with the given properties.
  static ObjectSchema object(
    Map<String, AckSchema> properties, {
    List<String> required = const [],
    bool additionalProperties = false,
  }) =>
      ObjectSchema(
        properties.cast(),
        required: required,
        additionalProperties: additionalProperties,
      );

  /// Creates a discriminated object schema for polymorphic validation.
  static DiscriminatedObjectSchema discriminated({
    required String discriminatorKey,
    required Map<String, ObjectSchema> schemas,
  }) =>
      DiscriminatedObjectSchema(
        discriminatorKey: discriminatorKey,
        schemas: schemas,
      );

  /// Creates a list schema with the given item schema.
  static ListSchema<T> list<T extends Object>(AckSchema<T> itemSchema) =>
      ListSchema(itemSchema);

  /// Creates an enum schema for validating enum values.
  static EnumSchema<T> enumValues<T extends Enum>(List<T> values) =>
      EnumSchema(values: values);

  /// Creates a schema that can be one of many types.
  static AnyOfSchema anyOf(List<AckSchema> schemas) => AnyOfSchema(schemas);
}
