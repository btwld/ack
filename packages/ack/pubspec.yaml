name: ack
description: A simple validation library for Dart
version: 0.2.0
repository: https://github.com/btwld/ack

environment:
  sdk: ^3.3.0

dependencies:
  meta: ^1.15.0

dev_dependencies:
  coverage: ^1.11.1
  dart_code_metrics_presets: ^2.22.0
  lints: ^5.0.0
  test: ^1.24.0
