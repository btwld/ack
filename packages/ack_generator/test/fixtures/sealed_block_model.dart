import 'package:ack/ack.dart';

part 'sealed_block_model.g.dart';

/// A sealed block class that demonstrates polymorphic model pattern with discrimination
@Schema(
  description: 'Base block class with polymorphic subclasses',
  discriminatedKey: 'type',
)
sealed class SealedBlock {
  /// The type field used for discrimination
  final String type;

  /// Optional alignment property
  final String? align;

  /// Optional flexibility property
  final int? flex;

  /// Whether the block is scrollable
  final bool? scrollable;

  const SealedBlock({
    required this.type,
    this.align,
    this.flex,
    this.scrollable,
  });

  /// Convert to a JSON map
  Map<String, dynamic> toJson();
}

/// Text block with markdown content
@Schema(
  description: 'A text block with markdown content',
  discriminatedValue: 'text',
)
class TextBlock extends SealedBlock {
  /// The markdown content for this block
  final String content;

  const TextBlock({
    super.align,
    super.flex,
    super.scrollable,
    required this.content,
  }) : super(type: 'text');

  @override
  Map<String, dynamic> toJson() {
    return {
      'type': type,
      'content': content,
      if (align != null) 'align': align,
      if (flex != null) 'flex': flex,
      if (scrollable != null) 'scrollable': scrollable,
    };
  }
}

/// Image block with image source
@Schema(
  description: 'An image block with image source and attributes',
  discriminatedValue: 'image',
)
class ImageBlock extends SealedBlock {
  /// The image path or URL
  final String src;

  /// Optional width
  final double? width;

  /// Optional height
  final double? height;

  /// Optional image fit mode
  final String? fit;

  const ImageBlock({
    super.align,
    super.flex,
    super.scrollable,
    required this.src,
    this.width,
    this.height,
    this.fit,
  }) : super(type: 'image');

  @override
  Map<String, dynamic> toJson() {
    return {
      'type': type,
      'src': src,
      if (width != null) 'width': width,
      if (height != null) 'height': height,
      if (fit != null) 'fit': fit,
      if (align != null) 'align': align,
      if (flex != null) 'flex': flex,
      if (scrollable != null) 'scrollable': scrollable,
    };
  }
}

/// Widget block with custom properties
@Schema(
  description: 'A widget block with custom properties',
  additionalProperties: true,
  additionalPropertiesField: 'properties',
  discriminatedValue: 'widget',
)
class WidgetBlock extends SealedBlock {
  /// The widget name
  final String name;

  /// Any additional custom properties
  final Map<String, dynamic> properties;

  const WidgetBlock({
    super.align,
    super.flex,
    super.scrollable,
    required this.name,
    this.properties = const {},
  }) : super(type: 'widget');

  @override
  Map<String, dynamic> toJson() {
    return {
      'type': type,
      'name': name,
      ...properties,
      if (align != null) 'align': align,
      if (flex != null) 'flex': flex,
      if (scrollable != null) 'scrollable': scrollable,
    };
  }
}
