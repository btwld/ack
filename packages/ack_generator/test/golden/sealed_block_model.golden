// GENERATED CODE - DO NOT MODIFY BY HAND

// ignore_for_file: unnecessary_this, unnecessary_new, unnecessary_const, prefer_collection_literals
// ignore_for_file: lines_longer_than_80_chars, unnecessary_null_checks, non_constant_identifier_names

part of 'sealed_block_model.dart';

// **************************************************************************
// AckSchemaGenerator
// **************************************************************************

/// Generated base schema for SealedBlock with inheritance support
/// Base block class with polymorphic subclasses
class SealedBlockSchema extends BaseSchema {
  SealedBlockSchema([Object? super.value = null]);

  static final DiscriminatedObjectSchema schema = Ack.discriminated(
    discriminatorKey: 'type',
    schemas: {
      'text': TextBlockSchema.schema,
      'image': ImageBlockSchema.schema,
      'widget': WidgetBlockSchema.schema,
    },
  );

  static final ObjectSchema baseSchema = Ack.object(
    {
      'type': Ack.string,
      'align': Ack.string.nullable(),
      'flex': Ack.int.nullable(),
      'scrollable': Ack.boolean.nullable(),
    },
    required: ['type'],
    additionalProperties: false,
  );

  /// Ensures this schema and its dependencies are registered
  static void ensureInitialize() {
    SchemaRegistry.register<SealedBlockSchema>(
      (data) => SealedBlockSchema(data),
    );
    TextBlockSchema.ensureInitialize();
    ImageBlockSchema.ensureInitialize();
    WidgetBlockSchema.ensureInitialize();
  }

  @override
  AckSchema getSchema() => schema;

  String get type => getValue<String>('type')!;

  String? get align => getValue<String>('align');

  int? get flex => getValue<int>('flex');

  bool? get scrollable => getValue<bool>('scrollable');

  R when<R>({
    required R Function(TextBlockSchema) text,
    required R Function(ImageBlockSchema) image,
    required R Function(WidgetBlockSchema) widget,
  }) =>
      switch (type) {
        'text' => text(TextBlockSchema(toMap())),
        'image' => image(ImageBlockSchema(toMap())),
        'widget' => widget(WidgetBlockSchema(toMap())),
        _ => throw StateError('Unknown sealedblock type: $type'),
      };

  R maybeWhen<R>({
    R Function(TextBlockSchema)? text,
    R Function(ImageBlockSchema)? image,
    R Function(WidgetBlockSchema)? widget,
    required R Function() orElse,
  }) =>
      switch (type) {
        'text' => text?.call(TextBlockSchema(toMap())) ?? orElse(),
        'image' => image?.call(ImageBlockSchema(toMap())) ?? orElse(),
        'widget' => widget?.call(WidgetBlockSchema(toMap())) ?? orElse(),
        _ => orElse(),
      };

  /// Convert the schema to a JSON Schema
  static Map<String, Object?> toJsonSchema() =>
      JsonSchemaConverter(schema: schema).toSchema();
}

/// Generated schema for TextBlock extending SealedBlockSchema
/// A text block with markdown content
class TextBlockSchema extends SealedBlockSchema {
  TextBlockSchema([Object? super.value = null]);

  static final ObjectSchema schema = SealedBlockSchema.baseSchema.extend(
    {'content': Ack.string},
    required: ['content'],
    additionalProperties: false,
  );

  /// Ensures this schema and its dependencies are registered
  static void ensureInitialize() {
    SchemaRegistry.register<TextBlockSchema>(
      (data) => TextBlockSchema(data),
    );
  }

  @override
  AckSchema getSchema() => schema;

  String get content => getValue<String>('content')!;

  /// Convert the schema to a JSON Schema
  static Map<String, Object?> toJsonSchema() =>
      JsonSchemaConverter(schema: schema).toSchema();
}

/// Generated schema for ImageBlock extending SealedBlockSchema
/// An image block with image source and attributes
class ImageBlockSchema extends SealedBlockSchema {
  ImageBlockSchema([Object? super.value = null]);

  static final ObjectSchema schema = SealedBlockSchema.baseSchema.extend(
    {
      'src': Ack.string,
      'width': Ack.double.nullable(),
      'height': Ack.double.nullable(),
      'fit': Ack.string.nullable()
    },
    required: ['src'],
    additionalProperties: false,
  );

  /// Ensures this schema and its dependencies are registered
  static void ensureInitialize() {
    SchemaRegistry.register<ImageBlockSchema>(
      (data) => ImageBlockSchema(data),
    );
  }

  @override
  AckSchema getSchema() => schema;

  String get src => getValue<String>('src')!;

  double? get width => getValue<double>('width');

  double? get height => getValue<double>('height');

  String? get fit => getValue<String>('fit');

  /// Convert the schema to a JSON Schema
  static Map<String, Object?> toJsonSchema() =>
      JsonSchemaConverter(schema: schema).toSchema();
}

/// Generated schema for WidgetBlock extending SealedBlockSchema
/// A widget block with custom properties
class WidgetBlockSchema extends SealedBlockSchema {
  WidgetBlockSchema([Object? super.value = null]);

  static final ObjectSchema schema = SealedBlockSchema.baseSchema.extend(
    {
      'name': Ack.string,
      'properties': Ack.object({}, additionalProperties: true)
    },
    required: ['name'],
    additionalProperties: false,
  );

  /// Ensures this schema and its dependencies are registered
  static void ensureInitialize() {
    SchemaRegistry.register<WidgetBlockSchema>(
      (data) => WidgetBlockSchema(data),
    );
  }

  @override
  AckSchema getSchema() => schema;

  String get name => getValue<String>('name')!;

  Map<String, dynamic> get properties =>
      getValue<Map<String, Object?>>('properties')!;

  /// Convert the schema to a JSON Schema
  static Map<String, Object?> toJsonSchema() =>
      JsonSchemaConverter(schema: schema).toSchema();
}
